docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,168 @@
+some not.  Any runtime behavior should be isolated to the kubelet
+and information about remapping should come from the CRI.
+
+### Detecting and Driving Private User Namespaces Behavior
+
+In order to Do The Right Thing in any solution the system must be able 
+to detect that private user namespaces are in use.  
+
+In Docker, this is a daemon setting and could be exposed via 
+the `info` endpoint.  However, in rkt this is a run parameter so it is
+unknown if private user namespaces will be used until the container launch
+command is created.  
+
+As a consistent solution to both driving any behavior to escape 
+user namespaces (see system defaulting below) and to also control
+the use of `--private-users` in the rkt `run` command a simple feature
+gate can be added to the Kubelet as was done in the [experimental
+support](https://github.com/kubernetes/kubernetes/pull/31169/files#diff-c5e4440d8576a91980ab36fe63556326).
+
+### Mounts
+
+The system must ensure that the remapped UID and GID have
+access to the correct directories under `/var/lib/kubelet/pods` in order
+to perform mounting.
+
+This may be accomplished by ensuring group access to the correct
+directories.  Information about the GID mappings should be conveyed
+to the Kubelet and storage code via the CRI.
+
+TODO - are GIDs being remapped in the rkt implementation?
+
+TODO - specific information on mount types - some support chmod and
+should be able to be changed to the remapped uids and some do not. This
+needs to be called out.
+
+### Escaping Private User Namespaces
+
+Some containers may use features which prohibit the use of private
+user namespaces.  If the Kubelet is aware that private user namespaces
+are in use it has the ability to modify the container's runtime 
+configuration when it detects a feature that is not compatible with
+private user namespaces.
+
+If the system is to default host user namespace options it would occur
+when:
+
+1.  setting any other namespace flag to host.  Not strictly
+required for net=host if not binding to low ports but if the system
+provides defaulting it covers the 80% case
+1.  using capabilities that aren't tied to any namespace
+    * MKNOD
+    * SYS_MODULE
+    * SYS_TIME
+1.  explicitly granted permissions to use a host volume and escape the private user namespace
+
+This logic was implemented in the [experimental
+support](https://github.com/kubernetes/kubernetes/pull/31169) as of
+Kubernetes 1.5.  
+
+Open Question: should we allow the usage of host user namespace as a 
+flag in addition to defaulting?  Ie. exposing `HostUserNamespace` as 
+a flag on the `SecurityContext` like we do with net, ipc, and pid?
+
+### Enforcing Private User Namespaces
+
+In some cases a cluster administrator may wish to ensure that a container
+is never defaulted to use the host user namespace.  This falls under
+the responsibility of cluster policy with `PodSecurityPolicy`.  
+
+`PodSecurityPolicy` has the ability to control using the host network,
```

Comment:
@euank - after thinking about this a bit more I realized that we already have the ability to control what would be defaulted and what wouldn't via the PSP structures already in place.  This has the disadvantage of not being obvious (ie. a `AlwaysUserPrivateUserNs` on the `SecurityContext`) but should already be working.  

How do you feel about that path?